pkg/ipmasq/ipmasq.go

```diff
@@ -0,0 +1,180 @@
+
+package ipmasq
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"strings"
+	"time"
+
+	"gopkg.in/yaml.v2"
+
+	"github.com/cilium/cilium/pkg/controller"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/maps/ipmasq"
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "ipmasq")
+)
+
+// ipnet is a wrapper type for net.IPNet to enable de-serialization of CIDRs
+type Ipnet net.IPNet
+
+func (c *Ipnet) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var str string
+	if err := unmarshal(&str); err != nil {
+		return err
+	}
+
+	ip, n, err := net.ParseCIDR(strings.Trim(str, `"`))
+	if err != nil {
+		return fmt.Errorf("Invalid CIDR %s: %s", str, err)
+	}
+	if ip.To4() == nil {
+		return fmt.Errorf("Invalid CIDR %s: only IPv4 is supported", str)
+	}
+
+	*c = Ipnet(*n)
+	return nil
+
+}
+
+// config represents the ip-masq-agent configuration file encoded as YAML
+type config struct {
+	NonMasqCIDRs []Ipnet `yaml:"nonMasqueradeCIDRs"`
+}
+
+// IPMasqMap is an interface describing methods for manipulating an ipmasq map
+type IPMasqMap interface {
+	Update(cidr net.IPNet) error
+	Delete(cidr net.IPNet) error
+	Dump() ([]net.IPNet, error)
+}
+
+// IPMasqAgent represents a state of the ip-masq-agent
+type IPMasqAgent struct {
+	configPath             string
+	nonMasqCIDRsFromConfig map[string]net.IPNet
+	nonMasqCIDRsInMap      map[string]net.IPNet
+	ipMasqMap              IPMasqMap
+}
+
+// Start starts the "ip-masq-agent" controller which is used to sync the ipmasq
+// BPF maps.
+func Start(configPath string, syncPeriod time.Duration) error {
+	return start(configPath, syncPeriod, &ipmasq.IPMasqBPFMap{}, controller.NewManager())
+}
+
+func start(configPath string, syncPeriod time.Duration,
+	ipMasqMap IPMasqMap, manager *controller.Manager) error {
+
+	a := &IPMasqAgent{
+		configPath:        configPath,
+		nonMasqCIDRsInMap: map[string]net.IPNet{},
+		ipMasqMap:         ipMasqMap,
+	}
+
+	if err := a.restore(); err != nil {
+		return err
+	}
+
+	manager.UpdateController("ip-masq-agent",
+		controller.ControllerParams{
+			DoFunc: func(ctx context.Context) error {
+				return a.Update()
+			},
+			RunInterval: syncPeriod,
+		},
+	)
+
+	return nil
+}
+
+// Update updates the ipmasq BPF map entries with ones from the config file.
+func (a *IPMasqAgent) Update() error {
+	if err := a.readConfig(); err != nil {
+		return err
+	}
+
+	for cidrStr, cidr := range a.nonMasqCIDRsInMap {
+		if _, ok := a.nonMasqCIDRsFromConfig[cidrStr]; !ok {
+			log.WithField(logfields.CIDR, cidrStr).Info("Removing CIDR")
+			a.ipMasqMap.Delete(cidr)
+			delete(a.nonMasqCIDRsInMap, cidrStr)
+		}
+	}
+
+	for cidrStr, cidr := range a.nonMasqCIDRsFromConfig {
+		if _, ok := a.nonMasqCIDRsInMap[cidrStr]; !ok {
+			log.WithField(logfields.CIDR, cidrStr).Info("Adding CIDR")
+			a.ipMasqMap.Update(cidr)
+			a.nonMasqCIDRsInMap[cidrStr] = cidr
```

Comment:
I wonder if it's worth collecting the number of updates that occur here and either add metrics and/or a monitor message to state that there was a change in ipmasq CIDRs. Thought for operations, no need to address here in this PR.

Reply:
Hmm, I'd expect that in majority of cases users will configure the CIDRs once. However, if it's not a case, then we could add a metric and maybe the monitor message.

Reply:
Yeah mainly thinking "if/when we hit an issue, how do I know that ipmasq CIDRs was/was not a probable cause". If they're rarely reconfigured then it might not be a big deal.